from django.conf import settings
<<<<<<< HEAD
from django.contrib.sessions.backends.base import (
    CreateError, SessionBase, UpdateError,
)
=======
from django.contrib.sessions.backends.base import CreateError, SessionBase, UpdateError
>>>>>>> 32e7214c1350fb4a88064960159c92207483e581
from django.core.cache import caches

KEY_PREFIX = "django.contrib.sessions.cache"


class SessionStore(SessionBase):
    """
    A cache-based session store.
    """
<<<<<<< HEAD
=======

>>>>>>> 32e7214c1350fb4a88064960159c92207483e581
    cache_key_prefix = KEY_PREFIX

    def __init__(self, session_key=None):
        self._cache = caches[settings.SESSION_CACHE_ALIAS]
        super().__init__(session_key)

    @property
    def cache_key(self):
        return self.cache_key_prefix + self._get_or_create_session_key()

    def load(self):
        try:
            session_data = self._cache.get(self.cache_key)
        except Exception:
            # Some backends (e.g. memcache) raise an exception on invalid
            # cache keys. If this happens, reset the session. See #17810.
            session_data = None
        if session_data is not None:
            return session_data
        self._session_key = None
        return {}

    def create(self):
        # Because a cache can fail silently (e.g. memcache), we don't know if
        # we are failing to create a new session because of a key collision or
        # because the cache is missing. So we try for a (large) number of times
        # and then raise an exception. That's the risk you shoulder if using
        # cache backing.
        for i in range(10000):
            self._session_key = self._get_new_session_key()
            try:
                self.save(must_create=True)
            except CreateError:
                continue
            self.modified = True
            return
        raise RuntimeError(
            "Unable to create a new session key. "
<<<<<<< HEAD
            "It is likely that the cache is unavailable.")
=======
            "It is likely that the cache is unavailable."
        )
>>>>>>> 32e7214c1350fb4a88064960159c92207483e581

    def save(self, must_create=False):
        if self.session_key is None:
            return self.create()
        if must_create:
            func = self._cache.add
        elif self._cache.get(self.cache_key) is not None:
            func = self._cache.set
        else:
            raise UpdateError
<<<<<<< HEAD
        result = func(self.cache_key,
                      self._get_session(no_load=must_create),
                      self.get_expiry_age())
=======
        result = func(
            self.cache_key,
            self._get_session(no_load=must_create),
            self.get_expiry_age(),
        )
>>>>>>> 32e7214c1350fb4a88064960159c92207483e581
        if must_create and not result:
            raise CreateError

    def exists(self, session_key):
<<<<<<< HEAD
        return bool(session_key) and (self.cache_key_prefix + session_key) in self._cache
=======
        return (
            bool(session_key) and (self.cache_key_prefix + session_key) in self._cache
        )
>>>>>>> 32e7214c1350fb4a88064960159c92207483e581

    def delete(self, session_key=None):
        if session_key is None:
            if self.session_key is None:
                return
            session_key = self.session_key
        self._cache.delete(self.cache_key_prefix + session_key)

    @classmethod
    def clear_expired(cls):
        pass
